coins: remove SetFresh method from CCoinsCacheEntry

andrewtoth opened the pull request:
There is plenty of test code that exercises the code path for FRESH-but-not-DIRTY coins in CCoinsViewCache. However, we can see there is only one place in production code where we call `SetFresh` that is not preceded by `SetDirty`. This is in `CCoinsViewCache::FetchCoin` and we can see that this is called if the coin retrieved from `base->GetCoin` is spent. The `base` in this case can be another `CCoinsViewCache` or a `CCoinsViewDB`. In `CCoinsViewCache::GetCoin` we can see that we do not return spent coins, and in `CCoinsViewDB` we don't ever store spent coins so there are none to return.

Thus, we can safely remove this dead code, and now we can see that there are never any calls to `SetFresh` not preceded by `SetDirty`. This PR removes this dead code and replaces `SetFresh` by passing a `fresh` boolean flag to `SetDirty`. This simplifies the logic of `CCoinsViewCache` and lets us remove test cases checking for FRESH-but-not-DIRTY coins. It removes the possibility of FRESH-but-not-DIRTY coins being called from any code path, so we can eliminate this state entirely and reduce the cognitive load of having to consider it.

This is pulled out from https://github.com/bitcoin/bitcoin/pull/30673 to make the change simpler.

andrewtoth committed:
test: remove FRESH-but-not-DIRTY state from coins_tests

The FRESH-but-not-DIRTY state cannot occur in production code.
There are never any calls to SetFresh not preceded by SetDirty.
We should remove these test cases to simplify the test code and
better align it with production code.

andrewtoth committed:
coins: remove SetFresh method from CCoinsCacheEntry

There are no instances where SetFresh is called without being
preceded by SetDirty. This replaces SetFresh with a boolean flag
passed to SetDirty to indicate whether the DIRTY coin should also
be set to FRESH. This removes the possibility of setting a
FRESH-but-not-DIRTY cached coin state.

andrewtoth committed:
coins: inline CCoinsCacheEntry::AddFlags into SetDirty

There is only one callsite now for AddFlags, so we can inline
it into SetDirty.

l0rinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,8 +146,10 @@ class CoinsViewBottom final : public CCoinsView
+        }
         return std::nullopt;
     }
 
```
We should be able to remove `HaveCoin` now, since
```C++
bool CCoinsView::HaveCoin(const COutPoint &outpoint) const
{
    return GetCoin(outpoint).has_value();
}
```
is more accurate (especially with the new extra assert)

andrewtoth replied:
Done.

l0rinc replied:
Isn't that the same as the parent in `CCoinsView`?

andrewtoth replied:
Ah right, removed.

l0rinc commented on src/coins.cpp:
```diff
@@ -51,10 +51,6 @@ CCoinsMap::iterator CCoinsViewCache::FetchCoin(const COutPoint &outpoint) const
         if (auto coin{base->GetCoin(outpoint)}) {
             ret->second.coin = std::move(*coin);
             cachedCoinsUsage += ret->second.coin.DynamicMemoryUsage();
-            if (ret->second.coin.IsSpent()) { // TODO GetCoin cannot return spent coins
```
Now that we've removed it we should at least add an `Assume(!ret->second.coin.IsSpent());` here - or even an assert, since if we're not sure about this we shouldn't do it in the first place

andrewtoth replied:
Done.

l0rinc commented on src/coins.h:
```diff
@@ -188,14 +177,14 @@ struct CCoinsCacheEntry
     bool IsFresh() const noexcept { return m_flags & FRESH; }
 
-    //! Only call Next when this entry is DIRTY, FRESH, or both
+    //! Only call Next when this entry is DIRTY or DIRTY-and-FRESH
     CoinsCachePair* Next() const noexcept
     {
         Assume(m_flags);
```
I understand if you think it may modify too many things here, but I'd prefer an actual `Assume` instead of the comment:
```suggestion
    CoinsCachePair* Next() const noexcept
    {
        Assume(IsDirty());
        return m_next;
    }
```

(same for `Prev`)

andrewtoth replied:
I don't see how that is different. `m_flags` being set implies `IsDirty()` now, so it is equivalent. I prefer not to change this.

l0rinc replied:
`Assume(m_flags);` would be true for fresh only - `Assume(IsDirty());` is stricter

andrewtoth replied:
It's not possible to set fresh only now.

l0rinc replied:
Exactly, hence my recommendation

l0rinc commented on src/coins.h:
```diff
@@ -121,21 +121,6 @@ struct CCoinsCacheEntry
-    //! Adding a flag requires a reference to the sentinel of the flagged pair linked list.
-    static void AddFlags(uint8_t flags, CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept
-    {
-        Assume(flags & (DIRTY | FRESH));
```
we might want to mention in the commit message why we can safely remove this now

l0rinc replied:
Ok, agree, please resolve the comment

l0rinc commented on src/coins.h:
```diff
@@ -169,7 +154,16 @@ struct CCoinsCacheEntry
+        }
+        Assume(pair.second.m_prev && pair.second.m_next);
+        pair.second.m_flags |= DIRTY;
+        if (fresh) pair.second.m_flags |= FRESH;
```
What should happen when `fresh` is `false`?
Shouldn't we assert something like:
```C++
if (fresh) {
    pair.second.m_flags |= FRESH; 
} else {
    assert(!(pair.second.m_flags & FRESH));
}

l0rinc replied:
Should we remove its freshness in that case?

l0rinc replied:
But we have prohibited one of the states here, we can document that with code now instead of comments. If we think it's a behavior change the PR is incorrect, isn't it?

l0rinc commented on src/test/coinscachepair_tests.cpp:
```diff
@@ -154,16 +154,16 @@ BOOST_AUTO_TEST_CASE(linked_list_set_state)
-    BOOST_CHECK(n2.second.IsFresh() && !n2.second.IsDirty());
+    // Check that setting DIRTY and FRESH on new node inserts it after n1
+    CCoinsCacheEntry::SetDirty(n2, sentinel, /*fresh=*/true);
+    BOOST_CHECK(n2.second.IsFresh() && n2.second.IsDirty());
```
nit: for consistency with other `SetDirty` calls we could swap the parameter order:
```suggestion
    BOOST_CHECK(n2.second.IsDirty() && n2.second.IsFresh());
```

----

Alternatively we could rename `IsFresh` to `IsDirtyAndFresh` and assert that if fresh is true, so is dirty - which would simplify this line to asserting just a single method.

l0rinc commented on src/test/fuzz/coins_view.cpp:
```diff
@@ -150,8 +150,7 @@ void TestCoinsView(FuzzedDataProvider& fuzzed_data_provider, CCoinsView& backend
                     auto it{coins_map.emplace(random_out_point, std::move(coins_cache_entry)).first};
-                    if (dirty) CCoinsCacheEntry::SetDirty(*it, sentinel);
-                    if (fresh) CCoinsCacheEntry::SetFresh(*it, sentinel);
+                    if (dirty) CCoinsCacheEntry::SetDirty(*it, sentinel, fresh);
```
definition of `fresh` should likely include `dirty` now:
```C++
const auto fresh{dirty && fuzzed_data_provider.ConsumeBool()};
```

andrewtoth replied:
Done.

l0rinc commented on src/coins.h:
```diff
@@ -102,47 +102,25 @@ using CoinsCachePair = std::pair<const COutPoint, CCoinsCacheEntry>;
-     * They are set in SetDirty, SetFresh, and unset in SetClean.
-     * A flagged entry is any entry that is either DIRTY, FRESH, or both.
+     * They are set in SetDirty and unset in SetClean.
+     * A flagged entry is any entry that is DIRTY or DIRTY and FRESH.
```
```suggestion
     * A flagged entry is any entry that is DIRTY or DIRTY-and-FRESH.
```
(the or/and combination is confusing otherwise)

andrewtoth replied:
Done.

l0rinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,8 +146,10 @@ class CoinsViewBottom final : public CCoinsView
     {
-        // TODO GetCoin shouldn't return spent coins
-        if (auto it = m_data.find(outpoint); it != m_data.end()) return it->second;
+        if (auto it = m_data.find(outpoint); it != m_data.end()) {
```
nit: for consistency with other overrides:
```suggestion
        if (auto it{m_data.find(outpoint)}; it != m_data.end()) {
```

l0rinc commented on src/coins.cpp:
```diff
@@ -321,8 +315,8 @@ void CCoinsViewCache::SanityCheck() const
-        // Only 5 combinations are possible.
-        assert(attr != 2 && attr != 4 && attr != 7);
+        // Only 4 combinations are possible.
+        assert(attr != 2 && attr != 4 && attr != 6 && attr != 7);
```
great, now that we have this, can we call `SanityCheck` in more tests? I understood that wasn't the case since we still had invalid states in fuzz/unit tests

andrewtoth replied:
We can't since we are still calling this with spent, dirty, and fresh entries. These would be deleted in production code. That can also be tackled in a follow-up.

l0rinc replied:
You did enable it for a few additional cases, that's already a win 👍

andrewtoth replied:
Err, I did but didn't test locally before pushing. I will have to revert. Tests are failing.

andrewtoth replied:
I can make a follow-up to remove spent fresh and dirty entries from the test cases as well. I don't think that should happen in this PR though to keep it focused.
The previous PR also did this.

l0rinc reviewed (commented):
Concept ACK

Very glad this is getting cleanup up, I found it worrying that a lot of tests were exercising invalid states in the first place - not the best foundation.
Quickly went over the changes, I find the small commits and excellent commit messages easy to follow.
I'll run a full IBD over this (maybe with some extra `SanityCheck` sprinkled around for extra measure) to make sure real usage doesn't trigger any invalid states we're not aware of.

l0rinc commented on src/coins.cpp:
```diff
@@ -203,11 +198,10 @@ bool CCoinsViewCache::BatchWrite(CoinsViewCacheCursor& cursor, const uint256 &ha
                     entry.coin = it->second.coin;
                 }
                 cachedCoinsUsage += entry.coin.DynamicMemoryUsage();
-                CCoinsCacheEntry::SetDirty(*itUs, m_sentinel);
```
note: the cursor contains dirty entries now, so `if (!it->second.IsDirty()) {` on line 186 is also a noop - could we add a TODO there to signal that we're aware of it but want to tackle it in a follow-up?

Lalaweazel commented:
' DQ at need  f cc 2iooo9 to be in work iiiiiiii iroieeieeieowwiereo W e

On Mon, 21 July 2025, 12:02 pm Andrew Toth, ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/coins.h
> <https://github.com/bitcoin/bitcoin/pull/33018#discussion_r2218067867>:
>
> >   * - spent, not FRESH, DIRTY (e.g. a coin is spent and spentness needs to be flushed to the parent)
>   */
>  struct CCoinsCacheEntry
>  {
>  private:
>      /**
>       * These are used to create a doubly linked list of flagged entries.
> -     * They are set in SetDirty, SetFresh, and unset in SetClean.
> -     * A flagged entry is any entry that is either DIRTY, FRESH, or both.
> +     * They are set in SetDirty and unset in SetClean.
> +     * A flagged entry is any entry that is DIRTY or DIRTY and FRESH.
>
> Done.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/33018#discussion_r2218067867>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BLCZ66TZ63CTQEG3FDJLXXD3JRC23AVCNFSM6AAAAACB46C4HCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTAMZWGMZDKOBZGM>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

sedited reviewed (commented):
Concept ACK

l0rinc commented on src/coins.h:
```diff
@@ -174,8 +167,10 @@ struct CCoinsCacheEntry
     }
 
-    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(DIRTY, pair, sentinel); }
-    static void SetFresh(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(FRESH, pair, sentinel); }
```
Had a [similar suggestion](https://github.com/bitcoin/bitcoin/pull/30673#discussion_r2206097629) in the old PR.
the new code is indeed more complicated. What if instead both `fresh` and `dirty` are parameters to a general `SetState` method?

andrewtoth committed:
coins: replace CCoinsCacheEntry::m_flags with m_fresh boolean

A cache entry is now always DIRTY if it is in the linked list.
We can use linked list membership instead of a flag to determine
dirtiness.
Thus, the only state we need to track is if an entry is FRESH.
This can be represented by a simple bool instead of a bitfield.

Co-authored-by: Lőrinc <pap.lorinc@gmail.com>

andrewtoth commented:
@sedited @l0rinc thank you both for your reviews! I rebased and updated with your suggestions.

l0rinc commented on src/coins.h:
Commit message main line of e9d793dc091c9a6359a4725ac1d0d734b983c6f4 has a typo

l0rinc commented on src/test/fuzz/coinscache_sim.cpp:
```diff
@@ -148,16 +146,13 @@ class CoinsViewBottom final : public CCoinsView
-        // TODO GetCoin shouldn't return spent coins
-        if (auto it = m_data.find(outpoint); it != m_data.end()) return it->second;
+        if (auto it{m_data.find(outpoint)}; it != m_data.end()) {
+            assert(!it->second.IsSpent());
```
let's be brave and make it an `Assert`

andrewtoth commented on src/coins.h:
```diff
@@ -102,103 +102,84 @@ using CoinsCachePair = std::pair<const COutPoint, CCoinsCacheEntry>;
 
-    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(DIRTY, pair, sentinel); }
-    static void SetFresh(CoinsCachePair& pair, CoinsCachePair& sentinel) noexcept { AddFlags(FRESH, pair, sentinel); }
+    static void SetDirty(CoinsCachePair& pair, CoinsCachePair& sentinel, bool fresh = false) noexcept
```
Added a docstring to this function and to `SetClean()`.

l0rinc reviewed (changes requested):
Re-reviewed and extracted a subset of the changes that aren't strictly related to freshness to https://github.com/bitcoin/bitcoin/pull/34207

Left a few comments about leftover commit messages and comments or symbols still referencing the old flags

l0rinc commented on src/coins.h:
```diff
@@ -191,26 +168,41 @@ struct CCoinsCacheEntry
-    //! Only call Next when this entry is DIRTY or DIRTY-and-FRESH
+    /**
+     * Dirty entries differ from the corresponding entry in the parent coins view. When flushing or syncing these coins
+     * must be written to the parent.
```
I don't usually mind long lines, but in this case we don't actually need them:
```suggestion
     * Dirty entries differ from the corresponding entry in the parent coins view.
     * When flushing or syncing these coins must be written to the parent.
```

l0rinc reviewed (changes requested):
code review ACK bb75e58f395364db22b2511f97b8479aa765c654


This is an excellent cleanup, removing even more of the impossible tested state, making the "freshness implies dirty" case super obvious.

I would prefer doing a few changes, but won't block it if others disagree.
I went over each commit again, tried to find counter-examples and imagine what it would look like after we get to #31132.

Some commits still mutate the code back and forth, I have suggested how we could simplify that.

The PR description needs some update:
* "This is pulled out from #30673 to make the change simpler" - not obvious to me if this is still the case now that #34207 was merged
* "However, we can see there is only one place in production code where we call SetFresh that is not preceded by SetDirty. This is in CCoinsViewCache::FetchCoin" - that's not true anymore, only `CCoinsViewCache::AddCoin()` and `CCoinsViewCache::BatchWrite()` have it, and both are preceded by `SetDirty()` - as claimed in the commit messages.

Eunovo commented on src/test/coins_tests.cpp:
https://github.com/bitcoin/bitcoin/pull/33018/commits/6f5dcf1f7780a05eedc818eabd2ff4005a5d23b3:

Tested this by commenting out each SetFresh/SetDirty call in coins.cpp and running coins_tests. Both SetFresh calls (in AddCoin, in BatchWrite) are exclusively caught by DIRTY_FRESH test cases; removing those test cases lets the bugs through undetected. The old FRESH-without-DIRTY test cases caught nothing that DIRTY_FRESH doesn't already cover. One gap found: EmplaceCoinInternalDANGER has no coins_tests coverage at all, but that's pre-existing, and https://github.com/bitcoin/bitcoin/pull/33512 already adds `EmplaceCoinInternalDANGER` to coin_tests

Eunovo reviewed (commented):
Concept ACK https://github.com/bitcoin/bitcoin/pull/33018/commits/bb75e58f395364db22b2511f97b8479aa765c654:

Left some comments
